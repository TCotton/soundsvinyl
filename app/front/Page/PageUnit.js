import React, { Component } from 'react'
import PropTypes from 'prop-types'
import ExtendedPropTypes from 'extended-proptypes'
import {
	thumbnailOne,
	thumbnailTwo,
	thumbnailThree,
	thumbnailFour,
	thumbnailFive
} from '../../alternative_default_thumbnails'
import { Link } from 'react-router-dom'
import './pageUnit.scss'

export default class PageUnit extends Component {
	static propTypes = {
		id: ExtendedPropTypes.mongoId.isRequired,
		slug: PropTypes.string.isRequired,
		subtitle: PropTypes.string.isRequired,
		thumbnailUrl: PropTypes.oneOfType([
			PropTypes.string,
			PropTypes.func,
		]).isRequired,
		title: PropTypes.string.isRequired
	}

	constructor (props) {
		super(props)

		this.handleOnError = this.handleOnError.bind(this)
	}

	handleOnError (event) {
		const myArray = [
			thumbnailOne,
			thumbnailTwo,
			thumbnailThree,
			thumbnailFour,
			thumbnailFive
		]
		const thumbnailDefault = myArray[Math.floor(Math.random() * myArray.length)]
		const randomNumber = Math.floor(Math.random() * 8) + 1

		event.target.className = `random${randomNumber}`;
		event.target.src = thumbnailDefault;
		event.target.alt = '';
	}

	render () {
		const { slug, id, thumbnailUrl, title, subtitle } = this.props

		return (
			<Link
				styleName='pageUnit'
				to={`/${slug}/${id}`}
			>
				<img
					alt={title}
					onError={this.handleOnError}
					src={thumbnailUrl}
				/>
				<span className='video-title'>
					{title}
				</span>
				<span className='video-subtitle'>
					{subtitle}
				</span>
			</Link>
		)
	}
}
