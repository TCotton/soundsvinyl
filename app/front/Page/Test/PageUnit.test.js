import React from 'react'
import PageUnit from '../PageUnit'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme';
import '../../../../enzymeConfig';
import { MemoryRouter as Router } from 'react-router-dom';
// import Observer from '@researchgate/react-intersection-observer';
// need to integrate @researchgate/react-intersection-observer';
describe.skip( 'Component', () => {
	let component

	const props = {
		id: '507f191e810c19729de860ea',
		slug: 'this-is-a-slug',
		subtitle: 'This is a subtitle',
		thumbnailUrl: 'https://thumbnailurl.com',
		title: 'This is a title'
	}

	beforeEach( () => {
		component = renderer.create(
			<Router>
				<PageUnit {...props} />
			</Router>
		)
	})

	describe( 'PageUnit', () => {
		it( 'should be defined', () => {
			expect( component ).toBeDefined()
		})

		it( 'snapshot', () => {

			const articleImage = renderer
				.create(
					<Router>
						<PageUnit {...props} />
					</Router>
				)
				.toJSON();

			expect( articleImage ).toMatchSnapshot();
		});
		// need to resolved props now '@researchgate/react-intersection-observer' is for IntersectionObserver
		it.skip( 'can be updated', () => {
			component.update(
				<Router>
					<PageUnit
						id='507f191e810c19729da860ea'
						slug='this-is-another-slug'
						subtitle='This is anothe subtitle'
						thumbnailUrl='https://thumbnailurl.co.uk'
						title='This is another title'
					/>
				</Router>
			)

			expect( component ).toBeDefined()
		})

		// '@researchgate/react-intersection-observer' now handles error in the handleIntersection method

		describe.skip( 'handling onError on image', () => {

			let component;
			let componentInstance;

			beforeAll( () => {
				component = mount(
					<Router>
						<PageUnit
							id='507f191e810c19729da860ea'
							slug='this-is-another-slug'
							subtitle='This is another subtitle'
							thumbnailUrl='not_a_URL'
							title='This is another title'
						/>
					</Router> ).find( PageUnit );

				componentInstance = component.instance();

			});

			it( 'Calls onError when passed bad src', () => {
				const event = {};
				event.target = {};

				componentInstance.handleOnError( event );

				expect( event.target.className ).toBeDefined( );
				expect( event.target.src ).toBeDefined( );
				expect( event.target.alt ).toBeDefined( );

				expect( event.target.className ).toEqual( expect.any( String ) );
				expect( event.target.src ).toEqual( expect.any( String ) );
				expect( event.target.alt ).toEqual( expect.any( String ) );

				expect( event.target.alt ).toHaveLength(0);
				expect( event.target.className.length ).toBeGreaterThan(0);
				expect( event.target.src.length ).toBeGreaterThan(0);
			});
		})
	})
})
