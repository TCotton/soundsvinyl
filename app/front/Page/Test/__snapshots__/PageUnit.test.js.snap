// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component PageUnit snapshot 1`] = `
<a
  className="pageUnit__pageUnit___mE_Gi"
  href="/this-is-a-slug/507f191e810c19729de860ea"
  onClick={[Function]}
>
  <img
    alt="This is a title"
    onError={[Function]}
    src="https://thumbnailurl.com"
  />
  <span
    className="video-title"
  >
    This is a title
  </span>
  <span
    className="video-subtitle"
  >
    This is a subtitle
  </span>
</a>
`;
